### PR TITLE
[ablation] Support for Keras custom models

### PR DESCRIPTION
**Pull request by ssheikholeslami**
Adds support for adding [custom Keras models](https://www.tensorflow.org/guide/keras/custom_layers_and_models) to an ablation study.

## API Changes
Adds a new method for adding a custom model generator:
- `ablationstudy.model.add_custom_model_generator(custom_model_generator, model_identifier)`

## Remaining Work
- [x] Update the example notebook with a custom model example
- [x] Fix the inline image issue
- [x] Black code formatting

### CODE REVIEW

File: maggy/ablation/ablationstudy.py [hunk 1]
```diff
@@ -203,9 +203,28 @@ def __init__(self):
...
+        # One (extra) trial will be generated per each item in the list.
+        self.custom_model_generators = [] 
+
     def set_base_model_generator(self, base_model_generator):
         self.base_model_generator = base_model_generator
-
+    
+    def add_custom_model_generator(self, custom_model_generator, model_identifier):
+        """
+        Add a custom model architecture generator, which will result in a single ablation trial.
```
Comment (moritzmeister) [inline maggy/ablation/ablationstudy.py]:
```suggestion
        Add a custom model architecture generator, which will result in a single ablation trial.

```
insert empty line after short description

File: maggy/ablation/ablationstudy.py [hunk 2]
```diff
@@ -203,9 +203,28 @@ def __init__(self):
...
-
+    
+    def add_custom_model_generator(self, custom_model_generator, model_identifier):
+        """
+        Add a custom model architecture generator, which will result in a single ablation trial. 
+        The `custom_model_generator` should be a Python callable, that once called should return a 
+        `keras.Model` object. This provides support for ablation study of 
+        arbitrarily complex models developed by extending `tf.Model` (for an example see 
+        [Writing custom layers and models with Keras](https://www.tensorflow.org/guide/keras/custom_layers_and_models#building_models)).
+        *Side effect*: This callable will be appended to `ablationstudy.Model.custom_model_generators`.
```
Comment (moritzmeister) [inline maggy/ablation/ablationstudy.py]: The user doesn't need to know about the internals. Desribe more in general that each custom generator will be a separate trial in the study.

File: maggy/ablation/ablationstudy.py [hunk 3]
```diff
@@ -203,9 +203,28 @@ def __init__(self):
...
+        
+        :param custom_model_generator: A Python callable that returns a `keras.Model`. This model generator will be
+        used as the model generation logic in the inner (training) loop of a single model ablation trial.
```
Comment (moritzmeister) [inline maggy/ablation/ablationstudy.py:219]: indent by 4

File: maggy/ablation/ablator/loco.py [hunk 4]
```diff
@@ -93,7 +94,10 @@ def decode(example_proto):
                     "Use 'tfrecord' or write your own custom dataset generator.",
                 )
 
-    def get_model_generator(self, layer_identifier=None):
+    def get_model_generator(self, layer_identifier=None, is_custom_model=False):
+
+        if is_custom_model:
```
Comment (moritzmeister) [inline maggy/ablation/ablator/loco.py]: doesn't do anything?

File: maggy/ablation/ablator/loco.py [hunk 5]
```diff
@@ -180,6 +185,16 @@ def initialize(self):
                     trial_type="ablation",
                 )
             )
+        
+        # 4 - generate ablation trials based on custom model generators
+
```
Comment (moritzmeister) [inline maggy/ablation/ablator/loco.py:199]: nitpick, remove empty line

File: maggy/ablation/ablator/loco.py [hunk 6]
```diff
@@ -203,7 +218,11 @@ def create_trial_dict(self, ablated_feature=None, layer_identifier=None):
         :rtype: dict
         """
 
+        #TODO modify trial dict structure: rename 'ablated_layer' to 'model configuration'
```
Comment (moritzmeister) [inline maggy/ablation/ablator/loco.py]: Are you going to address this in this PR?

File: maggy/ablation/ablator/loco.py [hunk 7]
```diff
@@ -213,12 +232,16 @@ def create_trial_dict(self, ablated_feature=None, layer_identifier=None):
             )
             trial_dict["ablated_feature"] = ablated_feature
 
-        if layer_identifier is None:
+        # 2 - determine the model generation logic
+        # 2.1 - no model ablation
+
```
Comment (moritzmeister) [inline maggy/ablation/ablator/loco.py:248]: nitpick, remove empty line

File: maggy/ablation/ablator/loco.py [hunk 8]
```diff
@@ -213,12 +232,16 @@ def create_trial_dict(self, ablated_feature=None, layer_identifier=None):
...
+        if layer_identifier is None and custom_model_generator is None :
```
Comment (moritzmeister) [inline maggy/ablation/ablator/loco.py]:
```suggestion
        if not layer_identifier and not custom_model_generator:
```
your call, whatever you like more, but remove space before the colon

File: maggy/ablation/ablator/loco.py [hunk 9]
```diff
@@ -213,12 +232,16 @@ def create_trial_dict(self, ablated_feature=None, layer_identifier=None):
...
             trial_dict[
                 "model_function"
             ] = self.ablation_study.model.base_model_generator
             trial_dict["ablated_layer"] = "None"
-        else:
+        # 2.2 - layer ablation based on base model generator
+        elif layer_identifier is not None and custom_model_generator is None:
```
Comment (moritzmeister) [inline maggy/ablation/ablator/loco.py:255]:
```suggestion
        elif layer_identifier and not custom_model_generator:
```
same as above, just syntactic sugar

File: maggy/ablation/ablator/loco.py [hunk 10]
```diff
@@ -231,5 +254,12 @@ def create_trial_dict(self, ablated_feature=None, layer_identifier=None):
                     trial_dict["ablated_layer"] = "Layers prefixed " + str(
                         list(layer_identifier)[0]
                     )
+        # 2.3 - model ablation based on a custom model generator
+        elif layer_identifier is None and custom_model_generator is not None:
```
Comment (moritzmeister) [inline maggy/ablation/ablator/loco.py:269]: see above

**Review by moritzmeister — CHANGES_REQUESTED** [review]
Need a few beauty changes. And definitely run black over it.

File: maggy/ablation/ablator/loco.py (hunk 4 above)
Comment (ssheikholeslami) [reply maggy/ablation/ablator/loco.py]: addressed in 3c5a47d .

File: maggy/ablation/ablator/loco.py (hunk 6 above)
Comment (ssheikholeslami) [reply maggy/ablation/ablator/loco.py]: No, this will go in a separate PR.

**Comment by ssheikholeslami** [comment]
Status update: I'm waiting for the new environment to test this PR, since we are getting those resource allocation errors unrelated to maggy. Once the tests are done, I will mark it as ready for review.

**Comment by ssheikholeslami** [comment]
@moritzmeister tested on Hopsworks 1.2.0.

File: maggy/ablation/ablationstudy.py [hunk 11]
```diff
@@ -341,6 +360,7 @@ def exclude_groups(self, *args, prefix=None):
     def print_all(self):
         """
         Prints all single layers that are included in the current ablation study configuration.
+
         """
```
Comment (moritzmeister) [inline maggy/ablation/ablationstudy.py]: Single line docstring can be on a single line
`"""Prints all single layers that are included in the current ablation study configuration."""`

File: maggy/ablation/ablationstudy.py [hunk 12]
```diff
@@ -352,6 +372,7 @@ def print_all(self):
     def print_all_groups(self):
         """
         Prints all layer groups that are included in the current ablation study configuration.
+
```
Comment (moritzmeister) [inline maggy/ablation/ablationstudy.py]: single line doc string

**Review by moritzmeister — APPROVED** [review]
:+1:

**Pull request merged**